compiler/rustc_lint/src/implicit_unsafe_autorefs.rs

```diff
@@ -0,0 +1,83 @@
+use crate::{LateContext, LateLintPass, LintContext};
+
+use rustc_errors::Applicability;
+use rustc_hir::{self as hir, Expr, ExprKind, UnOp};
+use rustc_middle::ty::adjustment::{Adjust, AutoBorrow};
+
+declare_lint! {
+    /// The `implicit_unsafe_autorefs` lint checks for implicitly taken references to dereferences of raw pointers.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// unsafe fn fun(ptr: *mut [u8]) -> *mut [u8] {
+    ///     addr_of_mut!((*ptr)[..16])
+    ///     //                 ^^^^^^ this calls `IndexMut::index_mut(&mut ..., ..16)`,
+    ///     //                        implicitly creating a reference
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// When working with raw pointers it's usually undesirable to create references,
+    /// since they inflict a lot of safety requirement. Unfortunately, it's possible
+    /// to take a reference to a dereferece of a raw pointer implitly, which inflicts
+    /// the usual reference requirements without you even knowing that.
```

Comment:
```suggestion
    /// When working with raw pointers it's usually undesirable to create references,
    /// since they inflict a lot of safety requirement. Unfortunately, it's possible
    /// to take a reference to a dereference of a raw pointer implicitly, which inflicts
    /// the usual reference requirements without you even knowing that.
```